### PR TITLE
Implement automatic differentiation engine

### PR DESCRIPTION
**Pull request by helloheyi**
Summary
This PR introduces an automatic differentiation engine using object-oriented C++. It supports both forward and backward a computational graph of operations e.g.+,/,-, etc.).

Key Components
  - Node classes: AddNode, SubNode, MulNode and etc 
  - Tape class: store and track nodes forward/backward passes
  - Examples in main.cpp

Features
  - Forward pass computes function value
  - Back-propagation computes gradients 
  - Finite difference function included for gradient and function verification
  - Edge case handling

Testing
  - Compared gradients with finite difference results.
  - Evaluated multiple test cases including:
  - Single-variable expressions
  - Multi-variable functions
  - Edge cases (e.g. x = -1, 0, large values)
  - Output validated via std::cout in main.cpp.

Next Steps
  - Add more complex node types (e.g., `SinNode`, `ExpNode`)
  - Handle higher-order or more complex multivariable functions

### CODE REVIEW

**Comment by christophercylai** [comment]
Hi, this is Chris. I am one of the interviewers. I am going to give comments and recommendations.
- I appreciate the comments, very helpful - despite a bit long
- The design of this project is over complicated: the expectation is minimum coding to satisfy the task at hand
- Please remove the taping/forward/backward
- Instead of making Node a parent class of Add (and others), can Add just be a class method of Node?
- Bonus: put everything into the main.cpp
- Bonus: use gcc to compile (https://code.visualstudio.com/docs/cpp/config-mingw), and give me the compilation command

File: main.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    auto final = tape.create<AddNode>(x3y, ln_xy);
+    // forward and backward pass
+    tape.forward();
+    tape.backward();
+    return std::vector<double>{final->getValue(), x->getGrad(), y->getGrad()};
+    
+}
+
+
+int main() {
```
Comment (christophercylai) [inline main.cpp]: - main function should take in the x_value from the command line (example: fte.exe 5)
- main function should write the results to a text file, without printing out the results (do print out the name of the text file that the results are written to)

File: main.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+ Parameters:
+ - x_0: Initial value of the variable x
+
+ Returns:
+ - A vector containing:
+   - f(x_0) (Function value)
+   - f'(x_0) (Gradient using reverse-mode autodiff)
+*/
+auto runGraphTest1(double x_0) {
```
Comment (christophercylai) [inline main.cpp]: move this into main()
also, there's no need to create so many variables - the node class should understand operators such as '+' and '-'

File: main.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ - x_0: Initial value of x
+ - y_0: Initial value of y
+
+ Returns:
+ - A vector containing:
+   - f(x_0, y_0) (Function value)
+   - df/dx (Gradient w.r.t. x)
+   - df/dy (Gradient w.r.t. y)
+*/
+auto runGraphTest2(double x_0, double y_0) {
```
Comment (christophercylai) [inline main.cpp]: move this into main()

File: main.cpp (hunk 1 above)
Comment (helloheyi) [reply main.cpp]: Thanks for the detailed feedback, Chris!


Updated `main.cpp` to:
- Accept `x` as a command-line argument (e.g., `./autoDiff 2`)
- Write the results (f(x), df/dx) to `results.txt` instead of printing them to stdout
- Print only the file name as confirmation, as requested

Let me know if there's anything else you'd like adjusted.

File: main.cpp (hunk 2 above)
Comment (helloheyi) [reply main.cpp]: - Updated main.cpp to build expression trees directly
- Each Node handles its own forward and backward pass logic

**Comment by helloheyi** [comment]
Hi Chris, thanks again for the thoughtful and direct feedback — much appreciated! 

I've incorporated your suggestions into two separate versions.

In Version 1 (the previous design)
  - I removed the Tape structure and recursive forward/backward infrastructure while keeping the object-oriented approach. 
  - Each operation (e.g., AddNode, PowNode) still inherits from a common Node base class, and gradients are computed via 
  - polymorphic method calls. While this version is more extensible, I recognize it introduces some extra complexity. 
  - Updated main.cpp to build expression trees
  -  Updated README

This version can be compiled g++ -std=c++14     main.cpp     nodes/ConstantNode.cpp     nodes/VariableNode.cpp     nodes/AddNode.cpp     nodes/SubNode.cpp     nodes/MulNode.cpp     nodes/DivNode.cpp     nodes/PowNode.cpp     nodes/LogNode.cpp     -o autoDiff
./autoDiff 2

In Version 2 (class_methods_of_Node/), 
I implemented a minimal design where each operation (Add, Sub, etc.) is defined directly as a class method of Node rather than using derived subclasses. 
  - This structure is more compact and clean, with the logic separated into main.cpp, node.cpp, and node.h. It aligns more 
     closely with your minimal code philosophy.
  - Updated main.cpp to build expression trees

This version can be compiled with g++ -std=c++14 main.cpp node.cpp -o class_methods, and run with ./class_methods 2

These executable files are uploaded as well. Let me know if you'd prefer an even more condensed version fully contained in main.cpp — happy to streamline further if needed!

File: main.cpp (hunk 3 above)
Comment (helloheyi) [reply main.cpp]: Updated main.cpp to build expression trees directly
Each Node handles its own forward and backward pass logic

**Comment by christophercylai** [comment]
Thanks! I am going to try this.
